### PR TITLE
chore: Add pre-commit configuration

### PR DESCRIPTION
**Pull request by PublisherName**
## Description

This pull request addes the `.pre-commit-config.yaml` file to include the latest versions of various pre-commit hooks. The changes ensure that our codebase adheres to the latest standards and best practices.

## Changes

- Added flake8 for code linting and PEP8 compliance
- Integrated black for consistent code formatting
- Configured isort to organize imports
- Added yamllint for YAML file validation
- Included check-added-large-files to prevent large files in commits

## Pre Commit Rules Implemented

- **Flake8**
  - No new rules are added to flake8.

- **Black**

  - For black formatting, an extend-exclude pattern has been added to exclude specific directories and files from formatting. This include .git, .venv, __pycache__ and migrations.

- **isort**
This isort configuration is set up to align with black formatting. New rules added are: 
   - Organizes imports into vertical sections, with each import on a separate line for readability.
   - Adds two blank lines after the import statements to maintain clarity and separation from the rest of the code.
   -  Combines as imports onto a single line where possible.
   -  Specifies directories to skip, including __pycache__, .venv, and migrations, to avoid sorting imports in automatically generated or unnecessary files.

- **check-added-large-files**
  - The check-added-large-files hook has been configured with its default setting, which flags any files larger than 500 kB. This helps prevent committing files that could significantly increase repository size, maintaining a manageable project footprint.

- **yamllint**
  -  line-length: Max 130 characters.
  -  braces/brackets: Max one space inside {} and [].
  - colons/commas: No spaces after : or ,.
  - comments: Disabled for flexibility.
  - document-start: --- start disabled.
  - empty-lines: Max two consecutive blank lines.

## How Has This Been Tested?

1. Ran `pre-commit install` to install the updated hooks.
2. Executed `pre-commit run --all-files` to ensure all hooks pass.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings

## Types of changes

- [x] Maintenance

### CODE REVIEW

**Comment by rex9840** [comment]
We are quite busy with deploying the server please have patience your pr is quite valuable we will gewt to you and your pr bit later @PublisherName

**Comment by rex9840** [comment]
Can you create a documentation for the pre-commit @PublisherName  so others know about  the pre-commits and the config to play around by the other maintainers and the conrtributors you can include it into the readme.md file

**Comment by PublisherName** [comment]
@rex9840 I noticed that there were no specific formatting or linting instructions in the README.md file, but I did find them documented in the CONTRIBUTING.md and style_guide.md. I've updated both of those files to include details on pre-commit configuration.

Please review the changes and let me know if there's anything else you'd like me to add or adjust.

File: .pre-commit-config.yaml
Comment (rex9840) [inline .pre-commit-config.yaml:16]: Just wanted to make somethign sure , so the logic is we want to format the code  with precommit rather than checking for the standars ?? @PublisherName

Comment (rex9840) [reply .pre-commit-config.yaml:16]: why did i asked the question is when running the pre-commit it fails but it  also format the code using black yml lint and the isort when testing next time it passes so its like magic first it fails and 2nd time it passes due to auto formatiing

Comment (rex9840) [reply .pre-commit-config.yaml:16]: if it is possible can you  also add the local check for the django check  ``python manage.py check`` duing the precomit  because some changes goes unnoticed sometimes  so the pre-commit can check wether the prtoject can run or not

Comment (PublisherName) [reply .pre-commit-config.yaml:16]: > Just wanted to make somethign sure , so the logic is we want to format the code with precommit rather than checking for the standars ?? @PublisherName

Yeah, exactly @rex9840 ! Running `python -m black .` and `python -m flake8 .` every time we make changes can be confusing, and we might even forget sometimes. With pre-commit configurations, the checks and formatting happen automatically whenever we use the `git commit` command. If any issues are detected, pre-commit takes care of them, ensuring the code meets the standards before it’s committed.

Comment (PublisherName) [reply .pre-commit-config.yaml:16]: > if it is possible can you also add the local check for the django check `python manage.py check` duing the precomit because some changes goes unnoticed sometimes so the pre-commit can check wether the prtoject can run or not

For sure! That’s a great idea. Adding python manage.py check to the pre-commit hooks will help catch any unnoticed issues before committing. Thank you for the suggestion!

Comment (PublisherName) [reply .pre-commit-config.yaml:16]: > why did i asked the question is when running the pre-commit it fails but it also format the code using black yml lint and the isort when testing next time it passes so its like magic first it fails and 2nd time it passes due to auto formatiing

Sorry, my bad! I forgot to fix a trailing whitespace in STYLE_GUIDE.md. I’ve committed the changes.

On the first run, pre-commit tries to fix any automatically solvable errors. If it succeeds, all checks will pass on the second run. But if any issues need manual fixing, they’ll remain until addressed.

**Comment by PublisherName** [comment]
@rex9840 Django configuration check has been added to pre-commit hooks.

File: .pre-commit-config.yaml
Comment (n1rjal) [reply .pre-commit-config.yaml:16]: Can you explain me what happens in case of an error, will the commit be rejected or will the commit be partially accepted ( even in rejection case, the PR will be accepted). 

I generaly like to put it in github actions as it make maintainers task easier for understanding style bugs and helping new contributors.

Comment (PublisherName) [reply .pre-commit-config.yaml:16]: If a pre-commit hook fails due to an error, the commit is fully rejected, no part of the commit goes through.

I haven’t configured precommits hooks in github actions for this PR yet, but if you’d like, I can add it to the workflow to automate these checks. Let me know if you'd like me to update this! @n1rjal

Comment (n1rjal) [reply .pre-commit-config.yaml:16]: I think isort is overkill here but I will give it a pass here. If you can remove it and also make a  worflow which will trigger on PR.

- Validation of code styles and pyflake checks on PR must be done during PR.

- We need the workflow to not change the code but only raise an error blocking merge. Later we can support contributors on how to improve it.

Comment (PublisherName) [reply .pre-commit-config.yaml:16]: Thank you for the feedback! I understand the preference for keeping the workflow lightweight, so I'll go ahead and remove **isort** and create a workflow that triggers on PRs. The new workflow will run code style validation and pyflake checks, blocking the merge if any issues are found. It won't make any automatic changes to the code.

I noticed there isn’t a PostgreSQL setup in the CI configuration. Does the Django Checks pass? Or am I missing something? Should I add a PostgreSQL setup to the workflow for the `python manage.py check` to run correctly?

https://github.com/GNOME-Nepal/site-server/blob/674c45f78f7c7dae61b7b9012ddb0835d04ff7e1/.github/workflows/CI.yml#L48-L50

Comment (n1rjal) [reply .pre-commit-config.yaml:16]:
```python3
 - name: Run Django system checks 
       run: | 
         python manage.py check || exit 1 
```

Yes, this only checks for syntax errors, and other similar kind of things, it is not a dedicated testing like pytest with some postgresql config. So postgresql is not needed here. Thankyou, waiting for the PR.

**Review by n1rjal — DISMISSED** [review]

**Comment by n1rjal** [comment]
Hi @PublisherName, I believe their is a merge conflict with `event/admin.py` please fix it.

**Review by n1rjal — CHANGES_REQUESTED** [review]
The workflow you had setup has failed, can you take a look at ot and also can you rebase and sync to our main and fix the merge conflicts

**Review by n1rjal — APPROVED** [review]

**Review by rex9840 — APPROVED** [review]

**Pull request merged**
